### ISSUE TITLE
NuGet downloads *way* too much for desktop projects

### ISSUE THREAD

**Issue by ericstj** [NuGet/Home#4734]
## Details about Problem

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe):
VS UI

NuGet version (x.x.x.xxx):
3.5.0

VS version (if appropriate):
Dev 14 update 2

OS version (i.e. win10 v1607 (14393.321)):
win10 v1607 (14393.693)

Worked before? If so, with which NuGet version:
Don't know

## Detailed repro steps so we can see the same problem

1.  Create a desktop console app, say net46

2. Install System.Net.Http 4.3.1 (though any package with multiple dependency groups should do).

Expected: NuGet only downloads the packages that are applicable and need to be installed.

Actual: NuGet downloads many packages for frameworks that aren't even applicable!

### Verbose Logs
https://gist.github.com/ericstj/3ef7e0373bab852ab46a4c46d151d1e5

You'll note that the package manager output shows that it only installs the following 5 packages:
System.Net.Http 4.3.1 
System.Security.Cryptography.Algorithms 4.3.0 
System.Security.Cryptography.Encoding 4.3.0 
System.Security.Cryptography.Primitives 4.3.0 
System.Security.Cryptography.X509Certificates 4.3.0 

However you will see that NuGet made **727** web requests (not including cached) covering over 30 different packages of various versions!

**Comment by emgarten** [comment]
@ericstj is there a way to tell beyond downloading the package and checking the files?

I would expect the package to contain a dependency group for net46 that contains only the correct dependencies. This is painful for the package author, but if the server did provide this info some scenarios would end up still needing to resolve these dependencies to check this again when the package was local since the feed info would be gone at that point.

**Comment by ericstj** [comment]
NuGet is downloading the package and (presumably) checking the files, but then it is still downloading things it knows it isn't going to use...  Every time it gets a package it can look at the nuspec and know deterministically the one set of dependencies it needs.  Why would it download more?

**Comment by emgarten** [comment]
For packages.config projects NuGet finds all possible versions of the package, then it finds all dependencies for the project target framework. So it is narrowing things down, but if an older or newer version is authored in a way that it looks like it could end up using more packages NuGet will get info from the feed for those also. 

The requests for RID packages in your output looks suspicious, packages.config wouldn't be reading runtime.json so it could only know about those through a nuspec.

**Comment by ericstj** [comment]
Crypto uses direct dependencies for RID specific packages so those are expected, though I wouldn't have ever expected NuGet to download them.

> For packages.config projects NuGet finds all possible versions of the package

Whoa, that's disturbing.  Especially since its looking at pre-release packages as well despite asking it to install stable.  Why does it do that if its only going to use the one you asked it to install?

I guess that could explain it.  I'll try using nuget.org exclusively and see if this is as bad.

**Comment by emgarten** [comment]
With PackageReferences the resolver doesn't go through all possible dependencies, so this will be improved once we move in the direction in the not so distant future.

The current behavior you are seeing is just for packages.config where resolution is stricter and making all http requests up front in parallel is sometimes beneficial.

**Comment by emgarten** [comment]
Closing as won't fix for packages.config, the solution to this problem is to use PackageReference.
